### PR TITLE
[eme] clearkey-mp4-playback-temporary-multikey-sequential-readyState.html playing count fix

### PR DESCRIPTION
**Pull request by cpearce**
Fix clearkey-mp4-playback-temporary-multikey-sequential-readyState.html so that it expects two "playing" events, as the specification calls for firing "playing" every time the readyState was HAVE_CURRENT_DATA or less and the new ready state is HAVE_FUTURE_DATA. The waiting for key algorithm requires us to drop the readyState to HAVE_METADATA, so we should expect to fire two "playing" events.

I have patches for Firefox in the works to make us pass this test with this fix.

### CODE REVIEW

**Comment by ddorwin** [comment]
I believe the text @cpearce is referring to is the following from https://html.spec.whatwg.org/multipage/embedded-content.html#ready-states:
>If the previous ready state was HAVE_CURRENT_DATA or less, and the new ready state is HAVE_FUTURE_DATA
&nbsp;&nbsp;The user agent must queue a task to fire a simple event named canplay at the element.
&nbsp;&nbsp;If the element's paused attribute is false, the user agent must notify about playing for the element.

Should we also have checks for the first step (`canplay`)?

File: encrypted-media/scripts/playback-temporary-multikey-sequential.js [hunk 1]
```diff
@@ -62,11 +63,12 @@ function runTest(config,qualifier) {
         }
 
         function onPlaying(event) {
-            assert_equals(_mediaKeySessions.length, 1, "Playback should start with a single key / session");
+            _playingCount++;
+            assert_equals(_mediaKeySessions.length, _playingCount, "Should get one 'playing' event per key / session played");
```
Comment (ddorwin) [inline encrypted-media/scripts/playback-temporary-multikey-sequential.js]: Should we also check that _playingCount is <= 2?

File: encrypted-media/scripts/playback-temporary-multikey-sequential.js [hunk 2]
```diff
@@ -62,11 +63,12 @@ function runTest(config,qualifier) {
...
         }
 
         function onTimeupdate(event) {
-            assert_false(_waitingForKey, "Should not continue playing whilst waiting for a key");
+            assert_true(!_waitingForKey || _video.readyState == _video.HAVE_METADATA, "Should not continue playing whilst waiting for a key");
```
Comment (ddorwin) [inline encrypted-media/scripts/playback-temporary-multikey-sequential.js]: Maybe explain why we should get a `timeupdate` when `_video.readyState == _video.HAVE_METADATA`. Is this at the beginning or on the transition due to waiting for a key?

**Comment by paulbrucecotton** [comment]
@cpearce :  Can you let us know if you are okay with David's comments?

**Comment by cpearce** [comment]
@paulbrucecotton In principle, yes I agree with those comments, but I'm not sure what readyState behaviour we're targeting until https://github.com/w3c/encrypted-media/issues/336, https://github.com/w3c/encrypted-media/issues/338, and https://github.com/w3c/encrypted-media/issues/339 are resolved. And this test is testing readyState behaviour.

File: encrypted-media/scripts/playback-temporary-multikey-sequential.js (hunk 1 above)
Comment (cpearce) [inline encrypted-media/scripts/playback-temporary-multikey-sequential.js]: Sure. Done.

File: encrypted-media/scripts/playback-temporary-multikey-sequential.js (hunk 2 above)
Comment (cpearce) [inline encrypted-media/scripts/playback-temporary-multikey-sequential.js]: Yes, we're required to fire 'timeupdate' when we transition from readyState HAVE_FUTURE_DATA or more to HAVE_CURRENT_DATA or less. I've added a comment.

File: encrypted-media/scripts/playback-temporary-multikey-sequential.js [hunk 3]
```diff
@@ -62,11 +64,24 @@ function runTest(config,qualifier) {
...
+            _canplayCount++;
+            assert_equals(_mediaKeySessions.length, _canplayCount, "Should get one 'canplay' event per key / session played");
+            assert_less_than_equal(_canplayCount, 2, "Should not get more than two 'canplay' events.");
         }
 
         function onTimeupdate(event) {
-            assert_false(_waitingForKey, "Should not continue playing whilst waiting for a key");
+            // We should not receive 'timeupdate' events due to playing while waiting for a key, except
+            // when we first start waiting for key we should change the readyState to HAVE_METADATA
+            // which will trigger the "If the previous ready state was HAVE_FUTURE_DATA or more, and
```
Comment (ddorwin) [inline encrypted-media/scripts/playback-temporary-multikey-sequential.js]: We'll need to update this once #338 is fixed.

File: encrypted-media/scripts/playback-temporary-multikey-sequential.js [hunk 4]
```diff
@@ -62,11 +64,24 @@ function runTest(config,qualifier) {
...
+            // the new ready state is HAVE_CURRENT_DATA or less" case of the readyState change
+            // algorithm which requires a "timeupdate" event be fired.
```
Comment (ddorwin) [inline encrypted-media/scripts/playback-temporary-multikey-sequential.js]: Don't you only want to see timeupdate once in that case?

It seems that the current logic is likely to always be true even if we get many timeupdate events while waiting.

**Review by ddorwin — COMMENTED** [review]
Thanks. Looks good except I have questions about one of the asserts.

File: encrypted-media/scripts/playback-temporary-multikey-sequential.js (hunk 3 above)
Comment (cpearce) [inline encrypted-media/scripts/playback-temporary-multikey-sequential.js]: Updated.

File: encrypted-media/scripts/playback-temporary-multikey-sequential.js (hunk 4 above)
Comment (cpearce) [inline encrypted-media/scripts/playback-temporary-multikey-sequential.js]: I've updated the test to count the number of timeupdates we get while waiting, and assert that we only get one. I've not been able to get the test to pass in Firefox yet.

**Comment by cpearce** [comment]
@ddorwin please re-review. Thanks!

File: encrypted-media/scripts/playback-temporary-multikey-sequential.js [hunk 5]
```diff
@@ -19,7 +19,10 @@ function runTest(config,qualifier) {
             _mediaKeys,
             _mediaKeySessions = [],
             _mediaSource,
-            _waitingForKey = false;
+            _waitingForKey = false,
+            _playingCount = 0,
+            _canplayCount = 0,
+            _timeupdateWhileWaiting = 0;
```
Comment (ddorwin) [inline encrypted-media/scripts/playback-temporary-multikey-sequential.js]: `...Count`?

File: encrypted-media/scripts/playback-temporary-multikey-sequential.js [hunk 6]
```diff
@@ -66,11 +69,27 @@ function runTest(config,qualifier) {
         }
 
         function onPlaying(event) {
-            assert_equals(_mediaKeySessions.length, 1, "Playback should start with a single key / session");
+            _playingCount++;
+            assert_equals(_mediaKeySessions.length, _playingCount, "Should get one 'playing' event per key / session played");
```
Comment (ddorwin) [inline encrypted-media/scripts/playback-temporary-multikey-sequential.js]: s/played/added/ ?

File: encrypted-media/scripts/playback-temporary-multikey-sequential.js [hunk 7]
```diff
@@ -66,11 +69,27 @@ function runTest(config,qualifier) {
...
+            assert_less_than_equal(_playingCount, 2, "Should not get more than two 'playing' events.");
+        }
+
+        function onCanPlay(event) {
+            _canplayCount++;
+            assert_equals(_mediaKeySessions.length, _canplayCount, "Should get one 'canplay' event per key / session played");
```
Comment (ddorwin) [inline encrypted-media/scripts/playback-temporary-multikey-sequential.js]: ditto: s/played/added/ ?

File: encrypted-media/scripts/playback-temporary-multikey-sequential.js [hunk 8]
```diff
@@ -66,11 +69,27 @@ function runTest(config,qualifier) {
...
 
         function onTimeupdate(event) {
-            assert_false(_waitingForKey, "Should not continue playing whilst waiting for a key");
+            // We should not receive 'timeupdate' events due to playing while waiting for a key, except
+            // when we first start waiting for key we should change the readyState to HAVE_CURRENT_DATA
+            // which will trigger the "If the previous ready state was HAVE_FUTURE_DATA or more, and
+            // the new ready state is HAVE_CURRENT_DATA or less" case of the readyState change
+            // algorithm which requires a "timeupdate" event be fired.
+             if (_waitingForKey) {
+                assert_equals(++_timeupdateWhileWaiting, 1, "Should only receive one timeupdate while waiting for key");
```
Comment (ddorwin) [inline encrypted-media/scripts/playback-temporary-multikey-sequential.js]: Before calling test.done(), we should assert_equals(_timeupdateWhileWaiting, 1) to ensure we did get that one instance. Currently, it's possible to get 0 and pass.

File: encrypted-media/scripts/playback-temporary-multikey-sequential.js [hunk 9]
```diff
@@ -66,11 +69,27 @@ function runTest(config,qualifier) {
...
+                assert_equals(_video.readyState, _video.HAVE_CURRENT_DATA, "Should not continue playing whilst waiting for a key");
```
Comment (ddorwin) [inline encrypted-media/scripts/playback-temporary-multikey-sequential.js]: Does this assert really have the meaning in the string? It seems that the logic above and expecting no more than one accomplishes that now.

**Review by ddorwin — APPROVED** [review]
This works, but I'd like to see the comments addressed. Feel free to get someone else to merge once you are satisfied.

File: encrypted-media/scripts/playback-temporary-multikey-sequential.js (hunk 9 above)
Comment (cpearce) [inline encrypted-media/scripts/playback-temporary-multikey-sequential.js]: Just to be clear, are you suggesting I remove the assert or change the string?

**Comment by cpearce** [comment]
Review comments addressed.

File: encrypted-media/scripts/playback-temporary-multikey-sequential.js [hunk 10]
```diff
@@ -87,12 +87,13 @@ function runTest(config,qualifier) {
...
              if (_waitingForKey) {
-                assert_equals(++_timeupdateWhileWaiting, 1, "Should only receive one timeupdate while waiting for key");
-                assert_equals(_video.readyState, _video.HAVE_CURRENT_DATA, "Should not continue playing whilst waiting for a key");
+                assert_equals(++_timeupdateWhileWaitingCount, 1, "Should only receive one timeupdate while waiting for key");
+                assert_equals(_video.readyState, _video.HAVE_CURRENT_DATA, "Video readyState should be HAVE_CURRENT_DATA while wating for key");
             }
 
             if (_video.currentTime > config.duration) {
                 assert_equals(_mediaKeySessions.length, config.initData.length, "It should require all keys to reach end of content");
+                assert_equals(_timeupdateWhileWaitingCount, 1, "Should have only received one timeupdate while waiting for key");
```
Comment (ddorwin) [inline encrypted-media/scripts/playback-temporary-multikey-sequential.js]: nit: ... received exactly one...

The check on line 90 checks no more than one. This checks that we did get that one.

**Review by ddorwin — APPROVED** [review]
Thanks. LGTM. One string could be clearer.

**Review by ddorwin — APPROVED** [review]

**Comment by ddorwin** [comment]
@cpearce, are you able to squash and merge? I was going to, but it's not clear what the commit message should be since there are many commits and it's not clear that the main ones are still accurate. If you have permission to merge, please do so with the commit message you want. Otherwise, post it here and I'll merge.

**Comment by cpearce** [comment]
I do not have write access, so I cannot perform a squash merge to master.

**Comment by ddorwin** [comment]
What do you want the commit message to be?

**Comment by cpearce** [comment]
"Ensure we receivethe expected number of canplay, playing and timeupdate events, and have the correct readyState, when playback waits for a key."

**Pull request merged**